cpp/src/arrow/compute/kernels/scalar_cast_temporal.cc

```diff
@@ -153,8 +155,12 @@ struct CastFunctor<
     const auto& in_type = checked_cast<const I&>(*batch[0].type());
     const auto& out_type = checked_cast<const O&>(*output->type);
 
-    // The units may be equal if the time zones are different. We might go to
-    // lengths to make this zero copy in the future but we leave it for now
+    if (I::type_id == Type::TIMESTAMP && in_type.unit() == out_type.unit()) {
+      std::shared_ptr<const ArrayData> array_data = input.ToArrayData();
+      output->SetMembers(*array_data);
+      arrow::compute::detail::PropagateNullsSpans(batch, output);
```

Comment:
I assumed (hoped) this would zero-copy the null bitmap. I suppose things are a little bit more subtle and I'll go read up a little bit.

Reply:
I think this is already done as part of `SetMembers`.  The null bitmap is just another buffer isn't it?

Reply:
Indeed. Removed `PropagateNullsSpans`. Oddly though this is still neither zero-copy or correct :).

Reply:
It seems the problem is that preallocation is still configured for the kernel.  So the executor preallocates a big output array and then iterates through views of the input.  When you change the output buffers you are only changing the view and not the big output array.

The other zero copy casts (using `AddZeroCopyCast` from scalar_cast_internal.cc) do not preallocate.  Ideally we would use AddZeroCopyCast for any same-unit-but-different-time-zone cast.  Unfortunately, in the general case, the output type of a kernel is resolved from the input type.  So when we are matching input types we don't know the output type.  So we can't add a matcher that is based on the output type.

One unpleasant but potential approach would be to disable preallocation on all timestamp->timestamp casts.  I think that would require changing the exec function to do the allocation though.  Perhaps there is some other technique we can use.

Reply:
Thanks for the explanation! I did try `AddZeroCopyCast` and hit the matcher issue. I'll take a look at the options you're proposing and report.

Reply:
I can start on disabling to preallocation. What are expected performance implications?
Alternatively we could error on or warn on same-unit-but-different-time-zone casts and provide a `SameTimeUnitCast` kernel that would be `AddZeroCopyCast` based.

Reply:
> I can start on disabling to preallocation. What are expected performance implications?

I'd expect them to be pretty minor.  When casting a chunked array it means we will do "# chunks" allocations instead of 1 allocation.  In Acero we never operate on chunked arrays so this is probably limited to people that are using the compute kernels directly.  Even then, this shouldn't be too big of a hit unless there were a lot of chunks.